Create a chat application using solidity and react on Avalanche

realnimish opened the pull request:
# Description*

I have added a tutorial to build a chat application using solidity and react on Avalanche's Fuji testnet

## Type of change*

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## This PR created for* - 

- [ ] DataHub Official Docs (Network Documentation, API docs etc.) changes/updates/fixes
- [x] Figment Learn Docs (Pathway, Existing Tutorials, New Tutorial etc.) changes/updates/fixes

# Checklist*:

- [x] I updated the `SUMMARY.md` file (FYI, It's needed only when you're creating a new file like submitting a new tutorial, pathway section)
- [x] I created one tutorial for this PR (one PR <=> one tutorial)
- [x] I checked and accept that my PR for new tutorial following the [**Tutorial Guidelines**](https://learn.figment.io/other/tutorial-guidelines) of this project
- [x] I checked and read the contribution guide before making this PR [**HERE**](https://learn.figment.io/figment-learn/contribute)  
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

yash-sharma1 commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,759 @@
+Following is the solidity contract which we will deploy on the network.
+<details>
+<summary>Contract</summary>
+
```
Instead of putting the code in `<summary>contract</summary>` tags put it just in a nomral code block to avoid the drop down function.

realnimish replied:
Fixed f59dd2fa0fe95dcea48de2e15ff1804267438b50

yash-sharma1 commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,759 @@
+<br><br>
+
+## Developing smart contract 
+We will have three types of user-defined data types, namely - `user`, `friend` & `message`. 
```
Instead of directly explaining the data types, please explain why you're writing this smart contract and what it will do? and then comes How? after that you can explain the data types and their functioning.

realnimish replied:
Added the explanation here 6a4c0a575e04f40b14cb684b4ab8b536ecdd6016

yash-sharma1 commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,759 @@
+[*Deploy a Smart Contract on Avalanche using Remix and MetaMask*](https://learn.figment.io/network-documentation/avalanche/tutorials/deploy-a-smart-contract-on-avalanche-using-remix-and-metamask). 
+
+>Note down the `contract address` and `ABI` generated in the above tutorial. They will be required in further steps.
+
```
Please take a error driven approach here where you can mention basic errors a developers/learner can face during the contract deployment process and some steps to troubleshoot the that problem. For reference you should [HERE](https://learn.figment.io/other/tutorial-guidelines/sample-tutorial-structure)

realnimish replied:
Added troubleshoot section 26b1b7f73793840964124826c9813a7c5c4fd6e7

yash-sharma1 commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,759 @@
+
+Make a `index.html` file in `public` folder of current directory and put the following html code.
+<details>
+<summary>index.html</summary>
```
Instead of putting the code in `summary` tags put all code part just in a nomral code block to avoid the drop down function (at all the steps)

realnimish replied:
Fixed here f59dd2fa0fe95dcea48de2e15ff1804267438b50

yash-sharma1 commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,759 @@
+<br>
+
+Make a new file named `abi.js` in the `src` folder. Export const variable named abi which stores the ABI you had generated earlier. Your abi.js file should look like this!
+
```
Here as an additional knoeledge point plz explain what is ABI?

realnimish replied:
Resolved here 8b61042fdf9c753a3e3deff945807b79f6fed4cb

yash-sharma1 commented:
@guillaumegaluz can you please this as well, I added my thoughts to it and request some changes.

yash-sharma1 reviewed (changes requested):
Please consider the above comments as requested changes.

guillaumegaluz commented:
Was this app developed from scratch or was it inspired by another existing tutorial or app?

realnimish commented:
> Was this app developed from scratch or was it inspired by another existing tutorial or app?

We have developed the application from scratch

yash-sharma1 commented:
@guillaumegaluz can you please review it once again and add your thoughts? Thanks

yash-sharma1 commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,802 @@
+                           {Messages}
+                        </div>
+                        {/*The form with send button and message input feilds*/}
+                        <div className="SendMessage"  style={{ borderTop:"2px solid black",position:"relative"bottom:"0px",padding:"10px 45px 0 45px",margin:"0 95px 0 0",
```
` <div className="SendMessage"  style={{ borderTop:"2px solid black",position:"relative"bottom:"0px",padding:"10px 45px 0 45px",margin:"0 95px 0 0",` is throwing error due because of missing `,`

It Should be 
` <div className="SendMessage"  style={{ borderTop:"2px solid black",position:"relative",bottom:"0px",padding:"10px 45px 0 45px",margin:"0 95px 0 0",`

yash-sharma1 commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
````diff
@@ -0,0 +1,802 @@
+```
+<br>
+
+Make a new file named `abi.js` in the `src` folder. Export const variable named abi which stores the ABI you had generated earlier. Your abi.js file should look like this!
````
You can write it as - 
Export const variable named abi which stores the ABI you had generated earlier. Add your abi const variable after `export const abi =` Your abi.js file should look like this!

yash-sharma1 commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,802 @@
+
+## About the Author(s)
+
+The tutorial was created by [Nimish Agrawal](https://github.com/realnimish) & [Sayan Kar](https://github.com/SayanKar). They are currently undergraduate students pursuing B.Tech in CSE [2018-2022] from Indian Institute of Information Technology, Guwahati. They both are highly passionate about the Blockchain Technology and are continuosly striving to learn more about the emerging protocols in the  ...
```
It should be short and informative so please write it as  -  
The tutorial was created by [Nimish Agrawal](https://github.com/realnimish) & [Sayan Kar](https://github.com/SayanKar). You can reach out to the authors on Figment Forum
- [Nimish Agrawal](https://community.figment.io/u/nimishagrawal100.in/)
- [Sayan Kar](Link to Sayan Kar's Figment Forum Profile)

yash-sharma1 commented:
In the end, while connecting to metamask I'm getting this error - 
![unknown](https://user-images.githubusercontent.com/44340561/119778345-05dba100-bee5-11eb-85b9-c9bd18ea4082.png)

realnimish commented:
> In the end, while connecting to metamask I'm getting this error -
> ![unknown](https://user-images.githubusercontent.com/44340561/119778345-05dba100-bee5-11eb-85b9-c9bd18ea4082.png)

You have most likely not added the contract address in the `index.js` file. Troubleshoot section already covers this issue under `Application crash` and to make the debugging more developer friendly we have added an alert message in the application itself in the following commit 18462e730023ba1f5817132ecf2523f946cb3697.

yash-sharma1 commented:
> > In the end, while connecting to metamask I'm getting this error -
> > ![unknown](https://user-images.githubusercontent.com/44340561/119778345-05dba100-bee5-11eb-85b9-c9bd18ea4082.png)
> 
> You have most likely not added the contract address in the `index.js` file. Troubleshoot section already covers this issue under `Application crash` and to make the debugging more developer friendly we have added an alert message in the application itself in the following commit [18462e7](https://github.com/figment-networks/datahub-learn/commit/18462e730023ba1f5817132ecf2523f946cb3697).

Nopes, I added the contract address in the `indes.js` file but still got the error - 
![Screenshot 2021-05-27 231001](https://user-images.githubusercontent.com/44340561/119872221-d9ee0900-bf40-11eb-8187-71484b34da54.png)

realnimish commented:
> > > In the end, while connecting to metamask I'm getting this error -
> > > ![unknown](https://user-images.githubusercontent.com/44340561/119778345-05dba100-bee5-11eb-85b9-c9bd18ea4082.png)
> > 
> > 
> > You have most likely not added the contract address in the `index.js` file. Troubleshoot section already covers this issue under `Application crash` and to make the debugging more developer friendly we have added an alert message in the application itself in the following commit [18462e7](https://github.com/figment-networks/datahub-learn/commit/18462e730023ba1f5817132ecf2523f946cb3697).
> 
> Nopes, I added the contract address in the `index.js` file but still got the error -
> ![Screenshot 2021-05-27 231001](https://user-images.githubusercontent.com/44340561/119872221-d9ee0900-bf40-11eb-8187-71484b34da54.png)

I think you have not deployed the contract on fuji testnet. There is no create transaction visible on [block explorer](https://testnet.avascan.info/blockchain/c/address/0xd8b934580fce35a11b58c6d73adee468a2833fa8/operations) for the given contract address.

Please make sure that you have selected `Injected Web3` while deploying from remix and have setup the fuji network configuration on Metamask as shown in the given [Tutorial - Deploy a Smart Contract on Avalanche using Remix and MetaMask](https://learn.figment.io/network-documentation/avalanche/tutorials/deploy-a-smart-contract-on-avalanche-using-remix-and-metamask) { Mentioned in the prerequisite of our tutorial }

yash-sharma1 commented:
> I think you have not deployed the contract on fuji testnet. There is no create transaction visible on [block explorer](https://testnet.avascan.info/blockchain/c/address/0xd8b934580fce35a11b58c6d73adee468a2833fa8/operations) for the given contract address.
> 
> Please make sure that you have selected `Injected Web3` while deploying from remix and have setup the fuji network configuration on Metamask as shown in the given [Tutorial - Deploy a Smart Contract on Avalanche using Remix and MetaMask](https://learn.figment.io/network-documentation/avalanche/tutorials/deploy-a-smart-contract-on-avalanche-using-remix-and-metamask) { Mentioned in the prerequisite of our tutorial }

Thanks, Let me check it once again

guillaumegaluz commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+
+* The second method registers _new user_ on the platform and finally the last method will return the username of the user if it exists.
+
+<br>
```
remove <br> here and everywhere else in the tutorial.

guillaumegaluz commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+
+* The first method as the name suggests is used to check if an user is registered with our application or not; which will be called in the `createAccount()` method to make sure duplicate user are not created and it will also be called later from other methods to check the user existence.
+
+* The second method registers _new user_ on the platform and finally the last method will return the username of the user if it exists.
```
I like how you explain those functions at a higher level by grouping them by theme but I would format it this way:

-------------------
We will define 3 functions:
- `checkUserExists(pubkey)` is used to check if a user is registered with our application or not. It will help make sure duplicate users are not created and it will also be called from other functions to check their existence.
- `createAccount(name)` registers a new user on the platform
- `getUsername(pubkey)` will return the username of the user if it exists.

guillaumegaluz commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+
+<br>
+
+Move out of the public folder and make a file named `Components.js` inside `src` folder and put the following code inside.
```
Instead of putting all the fucntional components into one file called `Components.jsx`, it's better to create a folder src/components` which will have one file per component: `AddNewChat.jsx`, `ChatCards.jsx`, etc

guillaumegaluz replied:
Use `.jsx` instead of `.js` if the file is exporting a component written with JSX

guillaumegaluz commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+}
+
+//This is a function which renders the friends in the friends list
+export function ChatCards(props){
```
Should this be `ChatCard` instead? I see only one card

guillaumegaluz commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+                    }}>
+                      Connect to MetaMask
+                    </Button>
+                    <div style={{display:props.showButton==="none"?"block":"none"}}>
```
add spaces after `:`

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+
+    // Login to metamask and check the if the user exists else creates one
+    async function login() {
+        /* Call connectToMetaMask */
```
This comment can be deleted.

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+    async function login() {
+        /* Call connectToMetaMask */
+        let res = await connectToMetamask();
+        if(res===true){
```
Proper spacing, please 😄 
> if( res === true ){

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+import ReactDom from "react-dom";
+import { useState, useEffect } from "react";
+import { Container, Row, Col, Card, Form, Button } from 'react-bootstrap';
+import { NavBar, ChatCards, Message, AddNewChat } from './Components';
```
The path you are requiring the components from would need to change, in accordance with @guillaumegaluz 's comment regarding a `components` subdirectory within the `src` directory..

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+
+export function App(props) {    
+
+    const contractAddress = ""; // Add the contract address inside the quotes
```
You can place the comment on the line above 😄

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+        if(res===true){
+            provider = new ethers.providers.Web3Provider(window.ethereum);
+            signer = provider.getSigner();
+            try{
```
> try {

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+            provider = new ethers.providers.Web3Provider(window.ethereum);
+            signer = provider.getSigner();
+            try{
+				const mycontract = new ethers.Contract(contractAddress, contractABI, signer);
```
Javascript variable names should be camelCase, so 'myContract'. This would require fixing it wherever it is referenced in the code.

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+				setMyPublicKey(address);		
+				setShowConnectButton("none");
+			} catch(err){
+				alert("You might not have set the contractAddress properly! Please check...")
```
> alert("contractAddress not set!")

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+				alert("You might not have set the contractAddress properly! Please check...")
+			}
+        }
+        else{
```
> else {

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+        }        
+    }
+
+    // Checks if the metamask connects 
```
> // Check if Metamask is available

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+
+    // Checks if the metamask connects 
+    async function connectToMetamask() {
+        /* Open Metamask window*/
```
This comment can be removed.

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+    // Checks if the metamask connects 
+    async function connectToMetamask() {
+        /* Open Metamask window*/
+        try{
```
> try {

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+		}
+    }
+
+    //Sends messsage to a User 
```
`// ` Space after comment slashes.

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+        if(!(activeChat && activeChat.publicKey)) return;
+        /* Send message */
+        const recieverAddress = activeChat.publicKey;
+        await myContract.sendMessage(recieverAddress,data);
```
Spaces after commas :smile:

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+                nickname = item.name;
+        });
+
+        /* Get message*/
```
Change to single line comment syntax.

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+        async function loadFriends(){
+            let friendList=[];
+
+            /* Get Friends*/
```
Change to single line comment syntax.

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+            let friendList=[];
+
+            /* Get Friends*/
+            try{
```
> try {

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+                data.map((item) => {
+                    friendList.push({"publicKey":item[0], "name":item[1]});
+                })
+            }catch(err){
```
Spacing is an important aspect of readability.

> } catch(err) {

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+            setFriends(friendList);
+        }
+        loadFriends();
+    },[myPublicKey,myContract]);
```
> }, [myPublicKey, myContract]);

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+        loadFriends();
+    },[myPublicKey,myContract]);
+
+    //Makes Cards for each Message
```
Space after `//`

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
````diff
@@ -0,0 +1,822 @@
+```
+<br>
+
+Make a new file named `abi.js` in the `src` folder. Export const variable named abi which stores the ABI you had generated earlier. Add your ABI after `export const abi = `. Your abi.js file should look like this!
````
Keep this section consistent with previous ones.

> Create a new file called `abi.js` in the `src` folder, and paste the following code :

pyr0gan replied:
Please explain what an ABI is.

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
````diff
@@ -0,0 +1,822 @@
+```
+* Visit [http://localhost:3000](http://localhost:3000) to interact with the app.
+
+* Don't forget to setup Metamask with `Fuji testnet` and also fund the account with Fuji c-chain test tokens for tx fees. Please refer to this tutorial on [Connecting Datahub to Metamask](https://learn.figment.io/network-documentation/avalanche/tutorials/connect-datahub-to-metamask).
````
Please expand on how to set up Metamask with the Fuji testnet. Linking to a separate tutorial is not sufficient.

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+![preview](https://github.com/realnimish/blockchain-chat-app/blob/main/public/UI.png?raw=true)
+
+## Conclusion
+Congratulatons!! You've successfully developed a distributed chatting application deployed on Avalanche's Fuji network and have also created a UI via which you can interact with the DApp.
```
> "Congratulations! We have successfully developed a decentralized chat application which can be deployed on Avalanche or other EVM-compatible blockchain. We also created a boilerplate React application to use as the frontend for our dApp."

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+
+* Check if your account has sufficient balance at [fuji block-explorer](https://testnet.avascan.info/). You can fund your address from the given [faucet](https://faucet.avax-test.network/) 
+
+* Make sure that you have selected the correct account on metamask if you have more than one account connected to the site.
```
It may be preferable to have an image to explain this.

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+<br>
+
+## What's Next
+The current DApp has very limited functionalities and we can improve it by adding features like deleting messages, blocking users, creating group(s) and optimising the DApp interaction cost with possible methods like max chat limit or using event log for short messages.
```
> This dApp has very limited functionality. We can improve it by adding functions to delete messages, block users, or create groups of friends. We could also optimize the dApp interaction cost with functions to limit the maximum number of messages, or possibly using event log for short messages.

pyr0gan reviewed (changes requested):
Please review my feedback and revise your tutorial 👍🏻 
We will go over your revisions next week. Enjoy the weekend!

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
````diff
@@ -149,93 +134,86 @@ contract Database {
+npm install --save ethers@5.1.4 react-bootstrap@1.5.2 bootstrap@4.6.0
 ```
 
 Now remove the contents of src and public folder
````
To clarify, I mean that you need to remove these instructions regarding the contents of /src and /public. It is not germane to the tutorial to remove the scaffolded files.

pyr0gan reviewed (commented):
Many of the issues are resolved & thank you for that, however there are still a few things outstanding. I have looked over the latest commit, and the formatting of some of the code blocks is still using a mix of tabs and spaces, which leads to some overly-indented code. We ask that you use 4 spaces per indent level.

We are approaching a final draft, just remember to get to the issues I pointed out regarding the information from other tutorials, as well as the 'Implementing' sections, moving the function descriptions to a bulleted list. See this comment by Guillaume : https://github.com/figment-networks/datahub-learn/pull/106/files/18462e730023ba1f5817132ecf2523f946cb3697..c432f1a75cc45d758a7041ebd31f567b5d9735fe#r641780098

realnimish commented:
Firstly I would like to thank all the reviewers for such a precise review (I think this PR is already way ahead of others in terms of number of interactions XD). The suggestions were to the point and it would surely have taken a lot of effort and maybe some strain on the eyes to identify the typos which we obviously missed in our inspection, sincere apologies for that. We are still working on the changes suggested and would request for the review as soon as those are finished. Thanks again!

guillaumegaluz commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -0,0 +1,822 @@
+
+## Prerequisites
+* Basic familiarity with Reactjs and Solidity
+* Should've completed [Deploy a Smart Contract on Avalanche using Remix and MetaMask](https://learn.figment.io/network-documentation/avalanche/tutorials/deploy-a-smart-contract-on-avalanche-using-remix-and-metamask) tutorial
```
@realnimish +1 on this. Deploying this is not trivial, and certainly nto something where we can assume that the learner will be able to easily use the other tutorial in the context of this one.

Although it sounds like we're duplicating some instructions, i think here it is the right thing to do - again - from the learner's perspective.

I would just copy and paste the deployement instructions in a new section here. Thanks!

guillaumegaluz commented:
> Firstly I would like to thank all the reviewers for such a precise review (I think this PR is already way ahead of others in terms of number of interactions XD). The suggestions were to the point and it would surely have taken a lot of effort and maybe some strain on the eyes to identify the typos which we obviously missed in our inspection, sincere apologies for that. We are still working on the changes suggested and would request for the review as soon as those are finished. Thanks again!

Thanks for your patience and thorough revisions @realnimish. We're trying to be more precise on reviews, not to be annoying, but to guarantee that the tutorials we are delivering to users will be of the highest quality. And please, no need to apologize!! 

There are only a few changes left here. As soon as they are done we will merge this and let the community know about this new tutorial!

SayanKar commented:
@pyr0gan @guillaumegaluz @yash-sharma1 Can you please review the PR, we have made all the requested changes.

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -18,7 +18,7 @@ Our chat dApp needs the basic functionality allowing users to connect with and s
 
 ### Account creation
```
I notice that the section headings are a little off. We should start with 1st level headings, so a single \# instead of \### . See the attached image: I made Introduction a 1st level heading, Prerequisites a 2nd level heading and Requirements a 3rd level heading, to illustrate the difference and how it renders on GitBook 😄 
![Screen Shot 2021-06-03](https://user-images.githubusercontent.com/2707197/120689455-5ca32680-c461-11eb-80ee-5fa6c85af995.png)

SayanKar replied:
@pyr0gan We have made a few changes and  have used the following format for heading levels.
# :- Tutorial Heading
## :- Section Heading
### :- Subsection heading 
It's not clear to me what changes you are requesting, can you please confirm ? If you are asking to make section headings 
1st level heading then the level will be the same as the tutorial heading.

pyr0gan replied:
Headings need to be H1, sub-headings are H2. Don't use H3 for anything that needs to show up on the table of contents is what I was getting at 😄 Perhaps the way I said it was confusing, apologies if it was unclear.

Introduction should be H1, it is a section heading. An example of a subheading would be in the Troubleshooting section, "Transaction Failure" this would be H2.
![Screen Shot 2021-06-07 at 10 21 41 AM](https://user-images.githubusercontent.com/2707197/121055212-4d321f00-c77a-11eb-80e2-1b482d47ed3f.png)

pyr0gan reviewed (commented):
Sorry, these changes are incorrect. Please refer to the sample tutorial structure document : [https://learn.figment.io/other/tutorial-guidelines/sample-tutorial-structure](https://learn.figment.io/other/tutorial-guidelines/sample-tutorial-structure)

SayanKar commented:
@pyr0gan I have made the changes you asked for, can you please check.

pyr0gan reviewed (commented):
Thank you for fixing the headings! Only a few more fixes to take care of and we should be able to launch today! 😄

realnimish commented:
> Thank you for fixing the headings! Only a few more fixes to take care of and we should be able to launch today! 😄

Cool! What all issues are remaining?

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -1,22 +1,22 @@
 
-## Introduction
+# Introduction
 Today we will build a decentralized chat application on Avalanche's Fuji test-network from scratch. The dApp will allow users to connect with other users and chat with them. We will develop our smart contract using Solidity which will be deployed on Avalanche's C-chain. It would have an easy-to-use UI developed using Reactjs. So Lets begin ...
```
Just a bit of polish needed here:

> Today we will build a decentralized chat application on Avalanche's Fuji test-network from scratch.
The dApp will allow users to connect with other people and chat with them. We will develop our smart contract using Solidity which will be deployed on Avalanche's C-chain. We will have a basic, easy-to-use UI developed using ReactJS. So, let us begin!

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -36,15 +36,15 @@ Here also we will define 3 functions :
 * The `readMessage()` function returns the chat history that has happened between the two users so far.
 
-### Data Collections
+## Data Collections
```
> User Data

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -26,7 +26,7 @@ We will define 3 functions :
 * The `getUsername(pubkey)` function will return the username of the given user if it exists.
 
-### Adding new friend
+## Adding new friend
```
> Adding friends

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -214,7 +214,7 @@ Approve the transaction on Metamask pop-up interface. Once our contract is deplo
 An Application Binary Interface (ABI) is a JSON object which stores the metadata about the methods of a contract like data type of input parameters, return data type & property of the method like payable, view, pure etc. You can learn more about the ABI from the [solidity documentation](https://docs.soliditylang.org/en/latest/abi-spec.html)  
```
GitHub won't let me select line 207 directly, but please make the following change :

> Navigate to the Solidity compiler Tab on the left side navigation bar and click the blue button to compile the `Database.sol` contract. Note down the `ABI` as it will be required in the next section.

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -214,7 +214,7 @@ Approve the transaction on Metamask pop-up interface. Once our contract is deplo
 {% endhint %}
 
-## Creating frontend using React
+# Creating frontend using React
```
> Creating a frontend in React

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -214,7 +214,7 @@ Approve the transaction on Metamask pop-up interface. Once our contract is deplo
-## Creating frontend using React
+# Creating frontend using React
 
 Now, we are going to create a react app and setup the frontend of the application.
```
> Now, we are going to create a React app scaffold and set up the frontend of the application.

pyr0gan commented on network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react.md:
```diff
@@ -78,7 +78,7 @@ Log in to MetaMask -> Click the Network drop-down -> Select Custom RPC
+## Deploy using Remix
 
 Open [Remix](https://remix.ethereum.org/) -> Select Solidity
 
```
It looks like the Solidity highlighting on GitBook isn't specified, so I think we can actually make the Solidity codeblock Javascript instead and it will look nicer with the syntax highlighting 😄

realnimish replied:
I previously used the Javascript code snippet for the very same reason :)
https://github.com/figment-networks/datahub-learn/pull/106#discussion_r641765520

pyr0gan replied:
Small detail, and Markdown rendering isn't universal :) Looks fine either way, some people prefer plain code and some people prefer the highlighting.

realnimish replied:
:raised_hands:

pyr0gan reviewed (commented):
Last little polish pass 👍🏻 Apply these and once another member of the team agrees with the changes, we should be able to merge and announce!

🚀

pyr0gan reviewed (approved):
Thanks for all your hard work on making this tutorial shine, @realnimish 😄 I think as long as my colleagues look it over once more and agree with the changes, we can merge.

realnimish commented:
> Thanks for all your hard work on making this tutorial shine, @realnimish smile I think as long as my colleagues look it over once more and agree with the changes, we can merge.

A big thanks to you guys too for all the support! I didn't imagined that I will get to learn this much in preparing a tutorial. Cheers!

cc: @yash-sharma1 @guillaumegaluz

yash-sharma1 commented:
Appreciate your work @realnimish & @SayanKar and your efforts in making small to small correction as per our feedback. Our main aim was to make this tutorial outstanding and improve as much as we all can and you guys did a great job in improving it. Looks good and approving the changes. 💯

pyr0gan commented:
@realnimish We've merged the tutorial and it is live! [View it here](https://learn.figment.io/network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react) 😄 

What's next? 🚀

realnimish commented:
I think the formatting near the "Application crash" needs to be updated!

realnimish commented:
> @realnimish We've merged the tutorial and it is live! [View it here](https://learn.figment.io/network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react) smile

Cool!

> What's next? rocket

We are looking into new ideas. Do you have any in mind that you would like us to work on?

yash-sharma1 commented:
> > @realnimish We've merged the tutorial and it is live! [View it here](https://learn.figment.io/network-documentation/avalanche/tutorials/create-a-chat-application-using-solidity-and-react) smile
> 
> Cool!
> 
> > What's next? rocket
> 
> We are looking into new ideas. Do you have any in mind that you would like us to work on?

That's great, you can check the available list of topics **[HERE](https://www.notion.so/Tutorial-Contribution-Ideas-d8ff9cdc32ca4b58838d81d07eab49bd)** also we would like to see some tutorial around DApp building and integration with DataHub so feel free to propose new ideas from your side as well.
